samples/drivers/led/pwm/src/main.c

```diff
@@ -80,31 +91,61 @@ static void run_led_test(const struct device *led_pwm, uint8_t led)
 	k_sleep(K_MSEC(1000));
 
 #if CONFIG_BLINK_DELAY_SHORT > 0
+	led_delay = (uint32_t)CONFIG_BLINK_DELAY_SHORT;
 	/* Start LED blinking (short cycle) */
-	err = led_blink(led_pwm, led, CONFIG_BLINK_DELAY_SHORT, CONFIG_BLINK_DELAY_SHORT);
+	err = led_blink(led_pwm, led, led_delay, led_delay);
 	if (err < 0) {
-		LOG_ERR("err=%d", err);
-		return;
+#if CONFIG_BLINK_DELAY_AUTO_FALLBACK && CONFIG_BLINK_DELAY_SHORT_LED_PERIOD_DIV > 0
+		LOG_INF("CONFIG_BLINK_DELAY_SHORT (%u ms) exceeds PWM capabilities for LED %u.",
+			led_delay, led);
+
+		led_delay = led_periods_ns[led] / 1000000 /
+			(uint32_t)CONFIG_BLINK_DELAY_SHORT_LED_PERIOD_DIV / 2;
+		/* Ensure minimum 1ms for visibility */
+		led_delay = MAX(1, led_delay);
+
+		LOG_INF("Auto-calculating delay: %u ms.", led_delay);
+
+		err = led_blink(led_pwm, led, led_delay, led_delay);
+#endif
+		if (err < 0) {
+			LOG_ERR("err=%d", err);
+			return;
+		}
 	}
 	LOG_INF("  Blinking "
-		"on: " STRINGIFY(CONFIG_BLINK_DELAY_SHORT) " msec, "
-		"off: " STRINGIFY(CONFIG_BLINK_DELAY_SHORT) " msec");
+		"on: %u msec, "
+		"off: %u msec", led_delay, led_delay);
 	k_sleep(K_MSEC(5000));
 #endif
 
 #if CONFIG_BLINK_DELAY_LONG > 0
+	led_delay = (uint32_t)CONFIG_BLINK_DELAY_LONG;
 	/* Start LED blinking (long cycle) */
-	err = led_blink(led_pwm, led, CONFIG_BLINK_DELAY_LONG, CONFIG_BLINK_DELAY_LONG);
+	err = led_blink(led_pwm, led, led_delay, led_delay);
 	if (err < 0) {
-		LOG_ERR("err=%d", err);
-		LOG_INF("  Cycle period not supported - "
-			"on: " STRINGIFY(CONFIG_BLINK_DELAY_LONG) "  msec, "
-			"off: " STRINGIFY(CONFIG_BLINK_DELAY_LONG) " msec");
-	} else {
-		LOG_INF("  Blinking "
-			"on: " STRINGIFY(CONFIG_BLINK_DELAY_LONG) " msec, "
-			"off: " STRINGIFY(CONFIG_BLINK_DELAY_LONG) " msec");
+#if CONFIG_BLINK_DELAY_AUTO_FALLBACK && CONFIG_BLINK_DELAY_LONG_LED_PERIOD_DIV > 0
+		LOG_INF("CONFIG_BLINK_DELAY_LONG (%u ms) exceeds PWM capabilities for LED %u.",
+			led_delay, led);
+		led_delay = led_periods_ns[led] / 1000000 /
```

Comment:
If you wrap these lines into a function they do not need to be duplicated, e.g.:

```c
int blink_delay_auto_fallback(const struct device *led_pwm, uint64_t period, uint32_t div)
{
		led_delay = period / 1000000 / div / 2;
		/* Ensure minimum 1ms for visibility */
		led_delay = MAX(1, led_delay);

		LOG_INF("Auto-calculated delay: %u ms.", led_delay);

		err = led_blink(led_pwm, led, led_delay, led_delay);
}
```